ouroboros-network-framework/src/Simulation/Network/Snocket.hs

```diff
@@ -199,13 +201,23 @@ data NetworkState m addr = NetworkState {
 
       -- | Registry of active connections.
       --
-      nsConnections       :: StrictTVar m (Map (NormalisedId addr) (Connection m addr)),
+      nsConnections       :: StrictTVar
+                              m
+                              (Map (NormalisedId addr) (Connection m addr)),
 
       -- | Get an unused ephemeral address.
       --
       nsNextEphemeralAddr :: AddressType -> STM m addr,
 
-      nsBearerInfo        :: LazySTM.TVar m (Script BearerInfo)
+      nsBearerInfo        :: LazySTM.TVar
+                              m
+                              (Script (LazySTM.STM m (LazySTM.TVar m (Script BearerInfo)))),
```

Comment:
Could you explain why we need script of scripts?

Is this #3378?  This will also not shrink well.  A better approach is to use a map `Map (ConnectionId peerAddr) (Script BearerInfo)`, this will allow to control how each connection is shrinking.  This map would be passed to `withSnocket`.  When we lookup the map and there's no script we should use `noAttenuation`.

Still I would like to hear your original idea, I am sure it is worth to consider.

Reply:
So `nsBearerInfo` is an infinite source (infinite because when it reaches the end of the script it returns always the same one) of `BearerInfo` TVars. Everytime a new connection arrives we check if an entry for it exists in the `nsAttenuationMap` if it doesn't, we reach `nsBearerInfo` for one and add an entry to the map, otherwise we just fetch the `TVar` and step the script.

Reply:
This is where everything happens.

https://github.com/input-output-hk/ouroboros-network/pull/3429/files/6bf811c01626b113f17fe2d23e24168ef72f5d0a..3d2352d22570002fca6c0da7ae13ac6a76668702#diff-9dc60e7be09dd9b4174122cea073cec55fb005ee452d64dafab6a12cfce6ed8bR746-R756

Reply:
I see.  This will not shrink well because if the simulation shrinks the connections it starts will start using different bearer infos anyway, but this wasn't the aim of this change, so ok.   I would suggest doing this differently though:

* keep `nsBearerInfo :: LazySTM.STM m (Script BearerInfo)`
* add `nsLastBearerInfo :: Script BearerInfo`.

Also this will be easier to translate into what I am eventually looking for: `Map ConnectionId (LazySTM.STM m (Script BearerInfo))` and a default `Script BearerInfo` for things that either don't exist in this map or reached the end of the script.

Reply:
btw, the link is stale, probably you forced pushed the branch after posting the link - that's a bit annoying from github.

Reply:
hmm, `nsAttenuationMap` is introduced below, so I am not sure why we need `nsBearerInfo` at all, just the default value for `BearerInfo` would suffice.

Reply:
fixed!